src/foam/nanos/auth/CachingAuthService.java

```diff
@@ -166,31 +165,39 @@ public void setX(X x) {
     super.setX(x);
 
     // TODO: Context is set on user permission cache for OM support, this can be removed when OM is removed from LRULinkListHashMap
-    userPermissionCache_.setX(getX());
+    permissionCache_.setX(getX());
   }
 
   public void start() throws Exception {
     Timer timer = new Timer(this.getClass().getSimpleName());
     timer.schedule(new AgencyTimerTask(getX(), this), INITIAL_TIMER_DELAY);
   }
 
-  protected Map<String, Boolean> getPermissionMap(User user) {
-    long userId = user == null ? -1 : user.getId();
-    Map<String, Boolean> map = userPermissionCache_.get(userId);
+  protected Map<String, Boolean> getPermissionMap(X x) {
+    String cacheKey = createCacheKey(x);
+
+    Map<String, Boolean> map = permissionCache_.get(cacheKey);
     if ( map == null ) {
       map = new ConcurrentHashMap<>();
-      userPermissionCache_.put(userId, map);
+      permissionCache_.put(cacheKey, map);
     }
 
     return map;
   }
 
-  private User getUserFromContext(X x) {
-    if ( x != null ) {
-      Subject subject = (Subject) x.get("subject");
-      return subject.getUser();
-    }
+  private String createCacheKey(X x) {
+    Subject subject = (Subject) x.get("subject");
+    User user = subject != null ? subject.getUser() : null;
+    User realUser = subject != null ? subject.getRealUser() : null;
+    Group group = (Group) x.get("group");
+
+    StringBuilder stringBuilder = new StringBuilder();
+    stringBuilder.append(user != null ? Long.toString(user.getId()) : "no-user");
```

Comment:
Doing Long.toString() still builds a garbage String. A more efficient method would be:
if ( user != null ) {
  sb.append(user.getId());
} else {
  sb.append("no-user");
}
Same on line 197.